### DISCUSSION TITLE
why are included relations not joined?

### DISCUSSION THREAD

**Discussion by ulo** [prisma/prisma#8840] [Q&A] [answered]
I am currently trying out Prisma and TypeORM to figure out which one to use for my data-intense web app.
I really like the developer experience with Prisma, but I am now wondering why querying related tables causes multiple select statements instead of a join?

The example is exactly the one described here: https://www.prisma.io/docs/guides/performance-and-optimization/query-optimization-performance#solving-n1-with-include

Prisma Client:
```ts
const usersWithPosts = await prisma.user.findMany({
  include: {
    posts: true,
  },
})
```

Resulting SQL:
```sql
SELECT "public"."User"."id", "public"."User"."email", "public"."User"."name" FROM "public"."User" WHERE 1=1 OFFSET $1
SELECT "public"."Post"."id", "public"."Post"."title", "public"."Post"."authorId" FROM "public"."Post" WHERE "public"."Post"."authorId" IN ($1,$2,$3,$4) OFFSET $5
```

Here the first select is querying the User table from the DB, and Prisma then takes the IDs to query the Post table. And from the query I can see that there have been 4 users in the database.

This does work and is also reasonably fast, however once I do have thousands of users the query gets ridiculously long, slower and eventually results in a "stack depth limit exceeded" error (of course I could increase `max_stack_depth` but this does not really solve the problem).

Is there a way to tell Prisma to use JOINs instead of these multiple SELECTs so that this findMany query would also work with e.g. 10000 users?

Would be great if someone has a comment or suggestion on this.
Many thanks!

**Comment by ryands17** [comment]
@ulo 👋 

We will be looking into joins and see if there's any noticeable change in performance so it would be great if you could follow these requests for more updates:
https://github.com/prisma/prisma/issues/5184
https://github.com/prisma/prisma/issues/4997
https://github.com/prisma/prisma/issues/2108

As a workaround, you can use the environment variable `QUERY_BATCH_SIZE` and set it to a value of **10000** if you're fetching that many amount of users. Enabling [`nApi`](https://www.prisma.io/docs/concepts/components/prisma-engines/query-engine#enable-the-node-api-n-api-preview) will also increase performance.

**Reply by ulo** [reply]
@ryands17 thanks for your reply! I had a look at your referenced issues and also at https://www.youtube.com/watch?v=RAoXdyI_PH4 and can understand Prisma's design decisions better now.

Indeed, I also did some benchmarks and for the simple case outlined above running multiple SELECTs is only twice as slow as the JOIN query, which is ok for me.

However, I did not really like increasing the `max_stack_depth` postgres setting to get it work. But the suggested `QUERY_BATCH_SIZE` is a much better solution. I can set it to 1000 and then easily query 10000s of users/posts. Even setting it to 100 works perfectly fine with only very minor performance costs.
I think it would be very helpful for Prisma to properly document the `QUERY_BATCH_SIZE` environment variable and maybe even set a sensible default value (e.g. 1000?) , so developers do not have to interpret the postgres error messages.

**Reply by ryands17** [reply]
Would it be possible for you to create an issue [here](https://github.com/prisma/docs/issues/new?assignees=&labels=docs&template=doc_request.md) so that we can persist this 🙂

**Comment by MCVitzz** [comment]
Are there any updates on this issue? I've ran into issues with large databases exceding max_stack_depth

**Reply by janpio** [reply]
This is not an issue but a discussion. If there is a problem that needs to be fixed, please open an actual issue and supply all the information the bug issue template asks for.

**Comment by jharrell** [comment, marked as answer]
Hi folks! Going to go ahead and close this discussion due to inactivity and previous answers. 

If you're interested in learning the theory behind multiple queries instead of joins, [check out this video]()

If you'd like more info on how to better your performance, our [query optimization docs](https://www.prisma.io/docs/guides/performance-and-optimization/query-optimization-performance) are a great resource. 

As always, we welcome any and all issues! https://github.com/prisma/prisma/issues/new/choose

**Reply by kamuiiiii** [reply]
It seems that the video link is not working properly.

**Reply by jharrell** [reply]
Sorry! I meant to link these two videos:

https://www.youtube.com/watch?v=RAoXdyI_PH4
https://www.youtube.com/watch?v=7oMfBGEdwsc
